### PR TITLE
[vulkan] Detect Vulkan / SPIR-V version & enabling required features

### PR DESCRIPTION
**Pull request by bobcao3**
Issue = #2612

### CODE REVIEW

File: taichi/backends/vulkan/codegen_vulkan.cpp [hunk 1]
```diff
@@ -49,6 +49,7 @@ class TaskCodegen : public IRVisitor {
  public:
   struct Params {
     OffloadedStmt *task_ir;
+    VkRuntime *runtime;
```
Comment (k-ye) [inline taichi/backends/vulkan/codegen_vulkan.cpp:52]: Is it possible to pass in just the `VulkanCapabilities`? The reason is that we should have a clean separation between the runtime stuff and codegen stuff. (In the future I'd like to completely separate them into different folders, e.g. `runtime/` and `codegen/`

File: taichi/backends/vulkan/spirv_ir_builder.cpp [hunk 2]
```diff
@@ -34,6 +25,10 @@ void IRBuilder::init_header(bool support_int8,
 
   // capability
   ib_.begin(spv::OpCapability).add(spv::CapabilityShader).commit(&header_);
+  // FIXME: What about devices don't support this?
+  ib_.begin(spv::OpCapability)
```
Comment (k-ye) [inline taichi/backends/vulkan/spirv_ir_builder.cpp:29]: That's a good question. Can we guard this somehow (at least by SPIR-V version?) We do need to put mobile platforms in mind here.

File: taichi/backends/vulkan/vulkan_api.cpp [hunk 3]
```diff
@@ -191,7 +192,8 @@ void ManagedVulkanDevice::create_instance(const Params &params) {
   app_info.applicationVersion = VK_MAKE_VERSION(1, 0, 0);
   app_info.pEngineName = "No Engine";
   app_info.engineVersion = VK_MAKE_VERSION(1, 0, 0);
-  app_info.apiVersion = params.api_version;  // important
+  app_info.apiVersion =
+      VK_API_VERSION_1_2;  // The highest version designed to use
```
Comment (k-ye) [inline taichi/backends/vulkan/vulkan_api.cpp:196]: Please still use `params.api_version` and update the API version there. I'd like to keep `vulkan_api` itself as platform-neutral as possible.

**Review by k-ye — COMMENTED** [review]
Thank!!

File: taichi/backends/vulkan/vulkan_api.cpp (hunk 3 above)
Comment (bobcao3) [reply taichi/backends/vulkan/vulkan_api.cpp:196]: Actually what does params.api_version do? Is it the max API version to use or is it the exact API version used?

Comment (k-ye) [reply taichi/backends/vulkan/vulkan_api.cpp:196]: >  is it the exact API version used

This

**Comment by bobcao3** [comment]
So we are targeting Vulkan 1.0 and use extensions for everything else?

**Comment by k-ye** [comment]
> So we are targeting Vulkan 1.0 and use extensions for everything else?

It's fine for Taichi to default at 1.2,  but ideally we should be able to support from 1.0 to 1.2. IIUC, https://github.com/taichi-dev/taichi/issues/2612 is a solution to help us achieve such support. @g1n0st ?

**Comment by g1n0st** [comment]
> > So we are targeting Vulkan 1.0 and use extensions for everything else?
> 
> It's fine for Taichi to default at 1.2, but ideally we should be able to support from 1.0 to 1.2. IIUC, #2612 is a solution to help us achieve such support. @g1n0st ?

I think this PR is exactly this solution. But Taichi may need a compile option or a runtime option to determine the Vulkan runtime and SPIR-V version?

**Comment by bobcao3** [comment]
The code in this PR does that. Vulkan 1.2 is requested, and according to
the standard either 1.1 or 1.2 will be reported back during detection. If
1.2 failed to init (which will only happen if only 1.0 is supported by the
loader), we will request 1.0 and the capabilities struct will have version
& related SPV version set to 1.0 & SPV 1.0. Variable pointer extension is
also detected in there, just that the codegen backend don't have a path to
generate code without that capability

**Comment by bobcao3** [comment]
Actually there's a slight problem with this PR rn but I don't see it really solved (or needs to be solved) soon. Under Vulkan 1.0 API, all devices that support Vulkan will support Vulkan 1.0, with various extensions turned on or off. Under Vulkan 1.1+ API, an individual physical device can report its API support, and this support is recorded into the VulkanCapabilities structure. So this means this structure is per-device, and this means the codegen will need to know the device it's on. I don't see an elegant solution (or need for it) for decoupling them until we have an unified device API.

**Comment by k-ye** [comment]
Thanks for the quick reply! I think we're on the same page now. I see why https://github.com/taichi-dev/taichi/pull/2615#discussion_r680293563 is confusing to you now. Here's what I think we can do:

1. rename that to `required_api_version`. However, make it an `std::optional`
2. When `required_api_version` is empty, do the auto-detection
3. Otherwise just use the `required_api_version`. This would allow our customers to ensure that Taichi does what they want.

**Review by g1n0st — APPROVED** [review]

File: taichi/backends/vulkan/codegen_vulkan.cpp (hunk 1 above)
Comment (k-ye) [reply taichi/backends/vulkan/codegen_vulkan.cpp:52]: Or at least leave a TODO here:

```suggestion
    // TODO: Pass in a VulkanCapabilities directly.
    VkRuntime *runtime;
```

File: taichi/backends/vulkan/runtime.h [hunk 4]
```diff
@@ -46,6 +48,10 @@ class VkRuntime {
 
   void synchronize();
 
+#ifdef TI_WITH_VULKAN
```
Comment (k-ye) [inline taichi/backends/vulkan/runtime.h:51]: Shouldn't really need to guard this: `VkRuntime` is implemented using the C++ Pimpl pattern.

**Review by k-ye — APPROVED** [review]
LGTM, thx!

**Pull request merged**